modin/data_management/dispatcher.py

```diff
@@ -0,0 +1,158 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+import os
+
+from modin import __execution_engine__ as execution_engine
+from modin import __partition_format__ as partition_format
+
+from modin.data_management import factories
+
+
+class StubIoEngine(object):
+    def __init__(self, factory_name=""):
+        self.factory_name = factory_name or "Unknown"
+
+    def __getattribute__(self, name):
+        factory_name = self.factory_name  # for closure to bind the value
+
+        def stub(*args, **kw):
+            raise NotImplementedError(
+                "Method {}.{} is not implemented".format(factory_name, name)
+            )
+
+        return stub
+
+
+class StubFactory(factories.BaseFactory):
+    """
+    A factory that does nothing more than raise NotImplementedError when any method is called.
+    Used for testing purposes.
+    """
+
+    io_cls = StubIoEngine()
+
+    @classmethod
+    def set_failing_name(cls, factory_name):
+        cls.io_cls = StubIoEngine(factory_name)
+        return cls
+
+
+class EngineDispatcher(object):
+    """
+    This is the 'ingestion' point which knows where to route the work
+    """
+
+    __engine = None
+
+    @classmethod
+    def _update_engine(cls):
+        if os.environ.get("MODIN_EXPERIMENTAL", "").title() == "True":
+            factory_fmt, experimental = "Experimental{}On{}Factory", True
+        else:
+            factory_fmt, experimental = "{}On{}Factory", False
+        factory_name = factory_fmt.format(partition_format, execution_engine)
+        try:
+            cls.__engine = getattr(factories, factory_name)
+        except AttributeError:
+            if not experimental:
+                # only allow missing factories in experimenal mode
+                raise
```

Comment:
Maybe to throw a special exception that will tell users which environment variables are incorrectly configured?

Reply:
I agree with @anmyachev. User may not be able to interpret the `AttributeError`.

We can provide the code for them to do that in the error message.

Reply:
Should be fixed by https://github.com/modin-project/modin/pull/1547/commits/6639e39a949ffee38bdf5f1f13200ea96400db59